README.md
[![discord](https://badgen.net/badge/icon/discord?icon=discord&label)](https://discord.gg/Qx2Ff3MByF)
![build](https://github.com/matyalatte/UE4-DDS-tools/actions/workflows/build.yml/badge.svg)
![test](https://github.com/matyalatte/UE4-DDS-tools/actions/workflows/test.yml/badge.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# UE4-DDS-Tools ver0.5.0

Texture modding tools for UE games.  
You can inject texture files (.dds, .tga, .hdr, etc.) into UE assets.  

## Features

- Inject any size textures into assets
- Extract textures from assets
- Check the UE versions of assets
- Convert textures with [texconv](https://github.com/microsoft/DirectXTex/wiki/Texconv)

## Supported UE versions

- UE5.0 ~ 5.1
- UE4.0 ~ 4.27
- FF7R
- Borderlands3

## Supported File Formats

This tool can convert textures between the following file formats.  

- .uasset (for texture assets)
- .dds
- .tga
- .hdr
- .bmp
- .jpg
- .png

> Note that it can not convert non-official DXGI formats (e.g. ASTC_4x4) from and to non-dds files.  
> It means you can not use .tga, .hdr, .bmp, .png, and .jpg for them.  

## Supported Pixel Formats

<details>
<summary>The List of supported pixel formats</summary>

- DXT1 (BC1)
- DXT3 (BC2)
- DXT5 (BC3)
- BC4 (ATI1)
- BC5 (ATI2)
- BC6H
- BC7
- A1
- A8
- G8 (R8)
- R8G8
- G16
- G16R16
- B8G8R8A8
- A2B10G10R10
- A16B16G16R16
- FloatRGB (FloatR11G11B10)
- FloatRGBA
- A32B32G32R32F
- B5G5R5A1_UNORM
- ASTC_4x4

</details>

> Note that Unreal Engine supports more [pixel formats](https://docs.unrealengine.com/5.0/en-US/API/Runtime/Core/EPixelFormat/).  
> You will get `Unsupported pixel format.` errors for them.  

## Supported Texture Classes

- Texture2D
- TextureCube
- LightMapTexture2D
- ShadowMapTexture2D
- Texture2DArray
- TextureCubeArray
- VolumeTexture

> Note that you can not use non-dds files for texture arrays and volume textures.

## Download

Download `UE4-DDS-tools*.zip` from [here](https://github.com/matyalatte/UE4-DDS-tools/releases)

## Getting Started

[How to Use · matyalatte/UE4-DDS-tools Wiki](https://github.com/matyalatte/UE4-DDS-Tools/wiki/How-to-Use)

## FAQ

[FAQ · matyalatte/UE4-DDS-Tools Wiki](https://github.com/matyalatte/UE4-DDS-Tools/wiki/FAQ)

## External Projects

### Texconv-Custom-DLL

[Texconv](https://github.com/microsoft/DirectXTex/wiki/Texconv)
is a texture converter developed by Microsoft.  
It's the best DDS converter as far as I know.  
And [Texconv-Custom-DLL](https://github.com/matyalatte/Texconv-Custom-DLL) is a cross-platform implementation I made.  
The official Texconv only supports Windows but you can use it on Unix systems.  

### Simple-Command-Runner

[Simple Command Runner](https://github.com/matyalatte/Simple-Command-Runner) is a GUI wrapper for executing commands.  
It can define a simple GUI with a json file.  

## License

* The files in this repository are licensed under [MIT license](https://github.com/matyalatte/UE4-DDS-Tools/blob/main/LICENSE).
* Released packeges use [Simple Command Runner](https://github.com/matyalatte/Simple-Command-Runner) for GUI. It is licensed under [wxWindows Library Licence](https://github.com/wxWidgets/wxWidgets/blob/master/docs/licence.txt).
* Released packeges use [Windows embeddable package](https://www.python.org/downloads/windows/) for python. It is licensed under [PSF license](https://docs.python.org/3/license.html).


changelog.txt
ver 0.5.0
- Supported Texture2DArray, TextureCubeArray, and VolumeTexture

ver 0.4.5
- Supported ShadowMapTexture2D
- Added filter options (Possible to use bicubic interpolation for mip generation)
- Updated GUI to v0.2.2 (GUI can save output path for each mode)
- Improved error massages for dds files
- Fixed a bug that the tool can't load uncompressed mipmaps correctly
- Fixed an error when loading non-zero values for FObjectExport.TemplateIndex
- Added tests

ver 0.4.4
- Supported UE4.0 ~ 4.9.
- Added "Skip non-texture assets" option.
- Fixed an error when loading assets that have multiple textures.
- Improved check mode (removed uneccessary versions from the check list)
- Removed parse errors for unsupported pixel formats
- Refined codes

ver 0.4.3
- Supported UE4.10 ~ 4.12
- Supported assets that have multiple texture objects
- Supported LightMapTexture2D
- Added a function to update PackageSource (use "MOD " for PackageSource)
- Updated texconv

ver 0.4.2
- Supported UE5.1
- Supported assets that have multiple texture objects (only for export mode)
- Fixed an error when assets have non-ascii characters.
- Fixed a bug that it won't update file size when updating name map.
- Refined codes

ver 0.4.1
- Supported more pixel formats
- Supported dds files that have no fourCC
- Added "force_uncompressed" option to inject mode

ver 0.4.0 (Introduced Texconv-Custom-DLL)
- Updated GUI (v0.1.2 to v0.2.1)
- Introduced tempfile for temporary files
- Introduced custom build for texconv (Texconv-Custom-DLL)
- Supported cubemaps
- Improved DDS handling

ver 0.3.3
- Supported UE5.0 assets
- Supported unversioned header
- Supported PF_ASTC_4X4

ver 0.3.2-hotfix
- Fixed a bug check mode won't work if the imported size is not the same as the actual size.


ver 0.3.2
- Added support for Borderlands3
- Removed legacy method (2-step injection with batch files)
- Improved folder method
- Minor updates for GUI

ver 0.3.1
- Added support for UE4.13, 4.14
- Fixed a bug check mode is not working
- Fixed an error when extracting uncompressed HDR textures (FloatRGBA)
- Minor updates for GUI

ver 0.3.0 (Introduced Texconv)
- Added support for tga, hdr, png, jpg, bmp. (You can inject non-dds files.)
- Added support for 4.16, 4.17
- Added --no_mipmaps option

ver 0.2.6
- Added support for 4.15, 4.20~4.24
- Added support for texconv to GUI

ver 0.2.5
- Added check mode (You can check which versions are the same format as your assets)
- Added check mode and export mode to GUI
- Fixed an error when loading assets corrupted by name table editing
- Modified comparison function
- Minor updates for GUI

ver 0.2.4
- Added support for UE4.25 games
- Added support for UE4.26 games
- Minor updates for GUI

ver 0.2.3
- Added support for GUI
- Made folder loader recursive
- Modified data structure
- Improved readability

ver 0.2.2
- Added support for Bloodstained

ver 0.2.1
- Added support for UE4.27 games
- Added support for cube maps
- Fixed an error when loading some assets.

ver 0.2.0 (Supported other UE versions)
- Added support for UE4.18 and UE4.19 games

ver 0.1.7
- Added support for .DDS
- Added feature to specify files in txt
- Fixed an error when loading some assets

ver 0.1.6
- Added support for BC7, DXT5/BC3, BC1
- Removed header files

ver 0.1.5
- Ignored sign when importing BC6H
- Added support for B8G8R8A8(sRGB)

ver 0.1.4
 - Added support for BC6H(unsigned)
 - Added support for any size mipmaps (no need to use power of 2 as height and width)

ver 0.1.3
 - Added export function
 - Bugfix

ver 0.1.2
 - Added support for ATI1 and ATI2

ver 0.1.1
 - Allow an folder as input.

ver 0.1.0
 - Initial version

src/main.py
'''Main file for UE4-DDS-Tools.'''
# std libs
import argparse
import json
import os
import time
from contextlib import redirect_stdout

# my scripts
from util import (compare, get_ext, get_temp_dir, flush_stdout,
                  get_file_list, get_base_folder, remove_quotes)
from unreal.uasset import Uasset
from directx.dds import DDS
from directx.dxgi_format import DXGI_FORMAT
from directx.texconv import Texconv, is_windows

TOOL_VERSION = '0.5.0'

# UE version: 4.0 ~ 5.1, ff7r, borderlands3
UE_VERSIONS = ['4.' + str(i) for i in range(28)] + ['5.' + str(i) for i in range(2)] + ['ff7r', 'borderlands3']

# Supported file extensions.
TEXTURES = ['dds', 'tga', 'hdr']
if is_windows():
    TEXTURES += ["bmp", "jpg", "png"]

# Supported image filters.
IMAGE_FILTERS = ["point", "linear", "cubic"]


def get_args():  # pragma: no cover
    parser = argparse.ArgumentParser()
    parser.add_argument('file', help='uasset, texture file, or folder')
    parser.add_argument('texture', nargs='?', help='texture file for injection mode.')
    parser.add_argument('--save_folder', default='output', type=str, help='output folder')
    parser.add_argument('--mode', default='inject', type=str,
                        help='valid, parse, inject, export, remove_mipmaps, check, convert, and copy are available.')
    parser.add_argument('--version', default=None, type=str,
                        help='UE version. it will overwrite the argment in config.json.')
    parser.add_argument('--export_as', default='dds', type=str,
                        help='format for export mode. dds, tga, png, jpg, and bmp are available.')
    parser.add_argument('--convert_to', default='tga', type=str,
                        help=("format for convert mode."
                              "tga, hdr, png, jpg, bmp, and DXGI formats (e.g. BC1_UNORM) are available."))
    parser.add_argument('--no_mipmaps', action='store_true',
                        help='force no mips to dds and uasset.')
    parser.add_argument('--force_uncompressed', action='store_true',
                        help='use uncompressed format for compressed texture assets.')
    parser.add_argument('--disable_tempfile', action='store_true',
                        help="store temporary files in the tool's directory.")
    parser.add_argument('--skip_non_texture', action='store_true',
                        help="disable errors about non-texture assets.")
    parser.add_argument('--image_filter', default='linear', type=str,
                        help=("image filter for mip generation."
                              " point, linear, and cubic are available."))
    parser.add_argument('--save_detected_version', action='store_true',
                        help="save detected version for batch file methods. this is an option for check mode.")
    return parser.parse_args()


def get_config():
    json_path = os.path.join(os.path.dirname(__file__), 'config.json')
    if not os.path.exists(json_path):
        return {}
    with open(json_path, encoding='utf-8') as f:
        return json.load(f)


def save_config(config):
    json_path = os.path.join(os.path.dirname(__file__), 'config.json')
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=4, ensure_ascii=False)


def parse(folder, file, args, texconv=None):
    '''Parse mode (parse dds or uasset)'''
    file = os.path.join(folder, file)
    if get_ext(file) == 'dds':
        DDS.load(file, verbose=True)
    else:
        Uasset(file, version=args.version, verbose=True)


def valid(folder, file, args, version=None, texconv=None):
    '''Valid mode (check if the tool can read and write a file correctly.)'''
    if version is None:
        version = args.version

    with get_temp_dir(disable_tempfile=args.disable_tempfile) as temp_dir:
        src_file = os.path.join(folder, file)
        new_file = os.path.join(temp_dir, file)

        if get_ext(file) == 'dds':
            # read and write dds
            dds = DDS.load(src_file)
            dds.save(new_file)

            # compare files
            compare(src_file, new_file)

        else:
            # read and write uasset
            asset = Uasset(src_file, version=version, verbose=True)
            uasset_name, uexp_name, ubulk_name = asset.get_all_file_path()
            asset.save(new_file, valid=True)
            new_uasset_name, new_uexp_name, new_ubulk_name = asset.get_all_file_path()

            # compare files
            compare(uasset_name, new_uasset_name)
            if new_uexp_name is not None:
                compare(uexp_name, new_uexp_name)
            if new_ubulk_name is not None:
                compare(ubulk_name, new_ubulk_name)


def search_texture_file(file_base, ext_list, index=None):
    """Sarch a texture file for injection mode."""
    if index is not None:
        file_base += f".{index}"
    for ext in ext_list:
        file = ".".join([file_base, ext])
        if os.path.exists(file):
            return file
    raise RuntimeError(f"Texture file not found. ({file_base})")


def inject(folder, file, args, texture_file=None, texconv=None):
    '''Inject mode (inject dds into the asset)'''

    # Read uasset
    uasset_file = os.path.join(folder, file)
    asset = Uasset(uasset_file, version=args.version)

    if not asset.has_textures():
        # Skip or raise an error for non-texture asset
        desc = f"(file: {uasset_file}, class: {asset.get_main_class_name()})"
        if args.skip_non_texture:
            print("Skipped a non-texture asset. " + desc)
            return
        raise RuntimeError("This uasset has no textures. " + desc)

    if texture_file is None:
        texture_file = args.texture
    file_base, ext = os.path.splitext(texture_file)
    ext = ext[1:].lower()
    if ext not in TEXTURES:
        raise RuntimeError(f'Unsupported texture format. ({ext})')

    textures = asset.get_texture_list()
    ext_list = [ext] + TEXTURES
    if len(textures) == 1:
        src_files = [search_texture_file(file_base, ext_list)]
    else:
        # Find other files for multiple textures
        splitted = file_base.split(".")
        if len(splitted) >= 2 and splitted[-1] == "0":
            file_base = ".".join(splitted[:-1])
        src_files = [search_texture_file(file_base, ext_list, index=i) for i in range(len(textures))]

    for tex, src in zip(textures, src_files):
        if args.force_uncompressed:
            tex.to_uncompressed()

        # Get a image as a DDS object
        if get_ext(src) == 'dds':
            dds = DDS.load(src)
        else:
            with get_temp_dir(disable_tempfile=args.disable_tempfile) as temp_dir:
                temp_dds = texconv.convert_to_dds(src, tex.dxgi_format,
                                                  out=temp_dir, export_as_cubemap=tex.is_cube,
                                                  no_mip=len(tex.mipmaps) <= 1 or args.no_mipmaps,
                                                  image_filter=args.image_filter,
                                                  allow_slow_codec=True, verbose=False)
                dds = DDS.load(temp_dds)

        # inject the DDS
        tex.inject_dds(dds)
        if args.no_mipmaps:
            tex.remove_mipmaps()
        flush_stdout()  # Need to flush to see the result for each asset

    # Write uasset
    asset.update_package_source(is_official=False)
    new_file = os.path.join(args.save_folder, file)
    asset.save(new_file)


def export(folder, file, args, texconv=None):
    '''Export mode (export uasset as dds)'''
    src_file = os.path.join(folder, file)
    new_file = os.path.join(args.save_folder, file)
    new_dir = os.path.dirname(new_file)

    asset = Uasset(src_file, version=args.version)

    if not asset.has_textures():
        # Skip or raise an error for non-texture asset
        desc = f"(file: {src_file}, class: {asset.get_main_class_name()})"
        if args.skip_non_texture:
            print("Skipped a non-texture asset. " + desc)
            return
        raise RuntimeError("This uasset has no textures. " + desc)

    textures = asset.get_texture_list()
    has_multi = len(textures) > 1
    for tex, i in zip(textures, range(len(textures))):
        if has_multi:
            # Add indices for multiple textures
            file_name = os.path.splitext(new_file)[0] + f'.{i}.dds'
        else:
            file_name = os.path.splitext(new_file)[0] + '.dds'

        if args.no_mipmaps:
            tex.remove_mipmaps()

        # Save texture
        dds = tex.get_dds()
        if args.export_as == 'dds':
            dds.save(file_name)
        else:
            if tex.is_3D or tex.is_array:
                print(
                    f"Warning: {args.export_as} is unsupported for {tex.get_texture_type()} textures."
                    " It'll be exported as dds."
                )
                dds.save(file_name)
            else:
                # Convert if the export format is not DDS
                with get_temp_dir(disable_tempfile=args.disable_tempfile) as temp_dir:
                    temp_dds = os.path.join(temp_dir, os.path.basename(file_name))
                    dds.save(temp_dds)
                    converted_file = texconv.convert_dds_to(temp_dds, out=new_dir, fmt=args.export_as, verbose=False)
                    print(f"convert to: {converted_file}")
        flush_stdout()  # Need to flush to see the result for each asset


def remove_mipmaps(folder, file, args, texconv=None):
    '''Remove mode (remove mipmaps from uasset)'''
    src_file = os.path.join(folder, file)
    new_file = os.path.join(args.save_folder, file)
    asset = Uasset(src_file, version=args.version)
    textures = asset.get_texture_list()
    for tex in textures:
        tex.remove_mipmaps()
    asset.save(new_file)


def copy(folder, file, args, texconv=None):
    """Copy mode (copy texture assets)"""
    # Read uasset
    uasset_file = os.path.join(folder, file)
    asset = Uasset(uasset_file, version=args.version)
    if not asset.has_textures():
        print("Skipped a non-texture asset.")
        return
    new_file = os.path.join(args.save_folder, file)
    asset.save(new_file)


# UE version for textures
UTEX_VERSIONS = [
    "5.1", "5.0",
    "4.26 ~ 4.27", "4.23 ~ 4.25", "4.20 ~ 4.22",
    "4.16 ~ 4.19", "4.15", "4.14", "4.12 ~ 4.13", "4.11", "4.10",
    "4.9", "4.8", "4.7", "4.4 ~ 4.6", "4.3", "4.0 ~ 4.2",
    "ff7r", "borderlands3"
]


def check_version(folder, file, args, texconv=None):
    '''Check mode (check file version)'''

    print('Running valid mode with each version...')
    passed_version = []
    for ver in UTEX_VERSIONS:
        try:
            # try to parse with null stdout
            with redirect_stdout(open(os.devnull, 'w')):
                valid(folder, file, args, ver.split(" ~ ")[0])
            print(f'  {(ver + " " * 11)[:11]}: Passed')
            passed_version.append(ver)
        except Exception:
            print(f'  {(ver + " " * 11)[:11]}: Failed')

    # Show the result.
    if len(passed_version) == 0:
        raise RuntimeError('Failed for all supported versions. You can not mod the asset with this tool.')
    elif len(passed_version) == 1 and ("~" not in passed_version[0]):
        print(f'The version is {passed_version[0]}.')
    else:
        s = f'{passed_version}'[1:-1].replace("'", "")
        print(f'Found some versions can handle the asset. ({s})')

    if args.save_detected_version:
        passed_version = [ver.split(" ~ ")[0] for ver in passed_version]
        common = list(set(passed_version) & set(args.version))
        if len(common) > 0:
            args.version = common
        else:
            args.version = passed_version


def convert(folder, file, args, texconv=None):
    '''Convert mode (convert texture files)'''
    src_file = os.path.join(folder, file)
    new_file = os.path.join(args.save_folder, file)

    if args.convert_to.lower() in TEXTURES[1:]:
        # not DDS
        ext = args.convert_to.lower()
    else:
        if not DXGI_FORMAT.is_valid_format(args.convert_to):
            raise RuntimeError(f"The specified format is undefined. ({args.convert_to})")
        # a DXGI format
        ext = "dds"

    new_file = os.path.splitext(new_file)[0] + "." + ext

    print(f"Converting {src_file} to {new_file}...")

    if ext == "dds":
        # image to dds
        texconv.convert_to_dds(src_file, DXGI_FORMAT[args.convert_to],
                               out=os.path.dirname(new_file), export_as_cubemap=False,
                               no_mip=args.no_mipmaps,
                               image_filter=args.image_filter,
                               allow_slow_codec=True, verbose=False)
    elif get_ext(file) == "dds":
        # dds to non-dds
        texconv.convert_dds_to(src_file, out=os.path.dirname(new_file), fmt=args.convert_to, verbose=False)
    else:
        # non-dds to non-dds
        texconv.convert_nondds(src_file, out=os.path.dirname(new_file), fmt=args.convert_to, verbose=False)


def main(args, config={}, texconv=None):
    # get config
    if (args.version is None) and ('version' in config) and (config['version'] is not None):
        args.version = config['version']

    if args.version is None:
        args.version = '4.27'

    if args.file.endswith(".txt"):
        # file path for batch file injection.
        # you can set an asset path with "echo some_path > some.txt"
        with open(args.file, 'r') as f:
            args.file = remove_quotes(f.readline())

    if args.mode == "check":
        if isinstance(args.version, str):
            args.version = [args.version]
    else:
        if isinstance(args.version, list):
            args.version = args.version[0]
        print(f'UE version: {args.version}')

    file = args.file
    texture_file = args.texture
    mode = args.mode

    print(f'Mode: {mode}')

    mode_functions = {
        'valid': valid,
        'inject': inject,
        'remove_mipmaps': remove_mipmaps,
        'parse': parse,
        'export': export,
        'check': check_version,
        "convert": convert,
        "copy": copy
    }

    # cehck args
    if os.path.isfile(args.save_folder):
        raise RuntimeError("Output path is not a folder.")
    if file == "":
        raise RuntimeError("Specify files.")
    if mode == 'inject' and (texture_file is None or texture_file == ""):
        raise RuntimeError("Specify texture file.")
    if mode not in mode_functions:
        raise RuntimeError(f'Unsupported mode. ({mode})')
    if mode != 'check' and args.version not in UE_VERSIONS:
        raise RuntimeError(f'Unsupported version. ({args.version})')
    if args.export_as not in ['tga', 'png', 'dds', 'jpg', 'bmp']:
        raise RuntimeError(f'Unsupported format to export. ({args.export_as})')
    if args.image_filter.lower() not in IMAGE_FILTERS:
        raise RuntimeError(f'Unsupported image filter. ({args.image_filter})')

    # load texconv
    if (mode == "export" and args.export_as != "dds") or mode in ["inject", "convert"]:
        if texconv is None:
            texconv = Texconv()

    func = mode_functions[mode]

    if os.path.isfile(file):
        folder = os.path.dirname(file)
        file = os.path.basename(file)
        func(folder, file, args, texconv=texconv)
    else:
        # folder method
        if mode == 'convert':
            ext_list = TEXTURES
        else:
            ext_list = ['uasset']

        folder, file_list = get_file_list(file, ext=ext_list, include_base=(mode != "inject"))

        if mode == 'inject':
            texture_folder = texture_file
            if not os.path.isdir(texture_folder):
                raise RuntimeError(
                    f'Specified a folder as uasset path. But texture path is not a folder. ({texture_folder})'
                )
            texture_file_list = [os.path.join(texture_folder, file[:-6] + TEXTURES[0]) for file in file_list]
            base_folder, folder = get_base_folder(folder)
            file_list = [os.path.join(folder, file) for file in file_list]
            for file, texture in zip(file_list, texture_file_list):
                func(base_folder, file, args, texture_file=texture, texconv=texconv)
                flush_stdout()
        else:
            for file in file_list:
                func(folder, file, args, texconv=texconv)
                flush_stdout()

    if mode == "check" and args.save_detected_version:
        print("Saved the detected version as src/config.json")
        if len(args.version) == 1:
            args.version = args.version[0]
        config['version'] = args.version
        save_config(config)


if __name__ == '__main__':  # pragma: no cover
    start_time = time.time()

    print(f'UE4 DDS Tools ver{TOOL_VERSION} by Matyalatte')

    args = get_args()
    config = get_config()
    main(args, config=config)

    if args.mode != "check":
        print(f'Success! Run time (s): {(time.time() - start_time)}')
